multiapi-engine/src/main/java/com/sngular/api/generator/plugin/asyncapi/AsyncApiGenerator.java

```diff
@@ -224,25 +232,53 @@ private Map<String, JsonNode> getAllSchemas(final FileLocation ymlParent, final
     final Map<String, JsonNode> totalSchemas = new HashMap<>();
     final List<JsonNode> referenceList = node.findValues(REF);
     referenceList.forEach(reference -> processReference(node, reference, ymlParent, totalSchemas, referenceList));
-    final List<JsonNode> messagesList = node.findValues(MESSAGES);
-    final List<JsonNode> schemasList = node.findValues(SCHEMAS);
-    schemasList.forEach(
-        schema -> schema.fields().forEachRemaining(fieldSchema -> totalSchemas.putIfAbsent((SCHEMAS + SLASH + fieldSchema.getKey()).toUpperCase(), fieldSchema.getValue())));
-    messagesList.forEach(message ->
-                             message.fields().forEachRemaining(fieldSchema -> {
-                               if (fieldSchema.getValue().has(PAYLOAD)) {
-                                 JsonNode payload = fieldSchema.getValue().get(PAYLOAD);
-                                 final String key = (MESSAGES + SLASH + fieldSchema.getKey()).toUpperCase();
-                                 if (payload.has(REF)) {
-                                   payload = payload.get(REF);
-                                 }
-                                 totalSchemas.putIfAbsent(key, payload);
-                               }
-                             })
-    );
+
+    if (node.has(COMPONENTS)) {
+      final List<JsonNode> schemasList = node.get(COMPONENTS).findValues(SCHEMAS);
+      schemasList.forEach(
+          schema -> schema.fields().forEachRemaining(
+              fieldSchema -> totalSchemas.putIfAbsent((SCHEMAS + SLASH + fieldSchema.getKey()).toUpperCase(), fieldSchema.getValue())
+          )
+      );
+
+      final List<JsonNode> messagesList = node.get(COMPONENTS).findValues(MESSAGES);
+      messagesList.forEach(
+          message -> getMessageSchemas(message, totalSchemas)
+      );
+    }
+
+    if (node.has(CHANNELS)) {
+      node.get(CHANNELS).forEach(
+          channel -> getChannelSchemas(channel, totalSchemas)
+      );
+    }
+
     return totalSchemas;
   }
 
+  private void getMessageSchemas(final JsonNode message, final Map<String, JsonNode> totalSchemas) {
+    if (message.has(PAYLOAD)) {
+      JsonNode payload = message.get(PAYLOAD);
+      if (!payload.has(REF)) {
+        final String key = (MESSAGES + SLASH + message.get(NAME).textValue()).toUpperCase();
+        if (payload.has(REF)) {
+          payload = payload.get(REF);
+        }
+        totalSchemas.putIfAbsent(key, payload);
+      }
+    }
+  }
+
+  private void getChannelSchemas(final JsonNode channel, final Map<String, JsonNode> totalSchemas) {
+    if (channel.has(PUBLISH) && channel.get(PUBLISH).has(MESSAGE)) {
```

Comment:
In Lines 273-274 and L.277-279 we have the same code (changing `SUBSCRIBE` instead of `PUBLISH`), in other projects we get the rule "refactor once it's repeated three times", so do as you wish.
According to the code, it seems possible that `channel` can contain both of them, so a possible solution could be a private function:
```
private void giveMeAGoodName (final JsonNode channel, final Map<String, JsonNode> totalSchemas, final String option) {
  if (channel.has(option) && channel.get(option).has(MESSAGE)) {
      getMessageSchemas(channel.get(option).get(MESSAGE), totalSchemas);
    }
}
```
Or there is another way if we put both options in an enum type or inside an array:
```
final Array<String> options = [PUBLISH, SUBSCRIBE];
options.forEach(option => {
  if (channel.has(option) && channel.get(option).has(MESSAGE)) {
      getMessageSchemas(channel.get(option).get(MESSAGE), totalSchemas);
    }
});
```
The last one would be great if we'll need to add more options.
Anyway, as I've said, I consider that your solution is valid.

Reply:
The last option is more concise than the original, so I'll probably go for that